src/tools/pkcs11-tool.c

```diff
@@ -2617,14 +2660,37 @@ static void decrypt_data(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
 			util_fatal("failed to open %s: %m", opt_output);
 	}
 
-	r = read(fd_in, in_buffer, sizeof(in_buffer));
-	in_len = r;
+	if (use_aead_buffer) {
+		in_len = lseek(fd_in, 0, SEEK_END);
+		aead_in_buffer = calloc(sizeof(CK_BYTE), in_len);
+		if (aead_in_buffer == NULL)
+			util_fatal("calloc failed: %m");
+		lseek(fd_in, 0, SEEK_SET);
+		r = read(fd_in, aead_in_buffer, in_len);
+	} else {
+		r = read(fd_in, in_buffer, sizeof(in_buffer));
+		in_len = r;
+	}
 
 	if (r < 0)
 		util_fatal("Cannot read from %s: %m", opt_input);
 
 	rv = CKR_CANCEL;
-	if (r < (int) sizeof(in_buffer)) {
+	if (use_aead_buffer) {
+		rv = p11->C_DecryptInit(session, &mech, key);
+		if (rv != CKR_OK)
+			p11_fatal("C_DecryptInit", rv);
+		if ((getCLASS(session, key) == CKO_PRIVATE_KEY) && getALWAYS_AUTHENTICATE(session, key))
+			login(session, CKU_CONTEXT_SPECIFIC);
+		out_len = sizeof(out_buffer);
```

Comment:
There is still a lot of duplicated code. In both of these conditions, this block is the same. Would it be possible to rewrite it to avoid the code duplication by placing this part before the newly introduced condition?